misc/init.d/cfengine3.in

```diff
@@ -348,10 +348,18 @@ cf_stop_daemon()
         fi
 
         pidfile_pid="$(cf_pidfile_pid "$pidfile")"
+        matching_pids="$(cf_get_matching_pids "$base_daemon")"
         if [ -n "$pidfile_pid" ]; then
-            pids="$pidfile_pid"
+            if ! echo "$matching_pids" | grep "$pidfile_pid" >/dev/null; then
+                # PID from the pid file not in the matching pids -> must be a
+                # leftover. Let's remove it.
+                rm -f "$pidfile"
+                pids="$matching_pids"
```

Comment:
This doesn't look line "clean the file and move on." as commit message says :-) I fail to imagine what can lead to a situation and what should be done when pidfile has one pid in it, and cf_get_matching_pids returns pid of some other process... But maybe I would do nothing in this case? I.e. `pids=""`.

Reply:
What do you mean? It removes the file and moves on to another iteration. The pidfile is clearly wrong if there's no such process. So I think it should be removed to prevent more confusion in other scripts etc. `$pids` is used as a list of processes to terminate. And if there are matching pids it means that there are processes that should be terminated. Or am I missing something?

Reply:
Right, I didn't realize that it happens at the end of the loop - I thought that you're going to do something with the $pids later. But now it's clear, you're right, sorry for the noise!

Reply:
> I fail to imagine what can lead to a situation

AFAICT that can happen if we have to use the KILL signal and kill the process forcibly in which case it has no chance to remove the pidfile. And a new process could be started right away with no pidfile created yet, for example.

Reply:
That's also why I only changed it in this one place.

Reply:
> if we have to use the KILL signal and kill the process forcibly

Hmm, I now see we remove the pidfile explicitly in that case. Anyway, I think this extra check is worth the 4 extra lines.